File: neoforge/src/main/java/dev/butterflysky/argus/neoforge/ArgusNeoForge.java
package dev.butterflysky.argus.neoforge;

import dev.butterflysky.argus.common.ArgusCore;
import net.minecraft.network.chat.Component;
import net.neoforged.bus.api.IEventBus;
import net.neoforged.bus.api.SubscribeEvent;
import net.neoforged.fml.common.Mod;
import net.neoforged.fml.ModContainer;
import net.neoforged.fml.event.lifecycle.FMLCommonSetupEvent;
import net.neoforged.neoforge.event.RegisterCommandsEvent;
import net.neoforged.neoforge.event.server.ServerStoppingEvent;
import net.neoforged.neoforge.common.NeoForge;
import net.neoforged.neoforge.event.entity.player.PlayerEvent;
import net.neoforged.neoforge.event.server.ServerStartedEvent;
import java.util.UUID;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/** NeoForge entrypoint wiring Argus shared logic. */
@Mod(ArgusNeoForge.MOD_ID)
public class ArgusNeoForge {
    public static final String MOD_ID = "argus";
    private static final Logger LOGGER = LoggerFactory.getLogger("argus-neoforge");
    private net.minecraft.server.MinecraftServer serverRef;

    public ArgusNeoForge(IEventBus modEventBus, ModContainer modContainer) {
        modEventBus.addListener(this::commonSetup);
        NeoForge.EVENT_BUS.register(this);
        NeoForge.EVENT_BUS.addListener(this::onRegisterCommands);
    }

    private void commonSetup(FMLCommonSetupEvent event) {
        ArgusCore.initializeJvm();
        ArgusCore.startDiscordJvm();
        ArgusCore.INSTANCE.registerMessenger((uuid, message) -> {
            if (serverRef != null) {
                var player = serverRef.getPlayerList().getPlayer(uuid);
                if (player != null) player.sendSystemMessage(Component.literal(message));
            }
            return kotlin.Unit.INSTANCE;
        });
        ArgusCore.INSTANCE.registerBanSync(
                (uuid, name, reason, until) -> {
                    if (serverRef != null) {
                        dev.butterflysky.argus.common.LoginIntrospection.ban(serverRef.getPlayerList(), uuid, name, reason, until);
                    }
                    return kotlin.Unit.INSTANCE;
                },
                uuid -> {
                    if (serverRef != null) {
                        dev.butterflysky.argus.common.LoginIntrospection.unban(serverRef.getPlayerList(), uuid);
                    }
                    return kotlin.Unit.INSTANCE;
                }
        );
    }

    private void onRegisterCommands(RegisterCommandsEvent event) {
        CommandReload.register(event.getDispatcher());
    }

    @SubscribeEvent
    public void onPlayerJoin(PlayerEvent.PlayerLoggedInEvent event) {
        var player = (net.minecraft.server.level.ServerPlayer) event.getEntity();
        var server = player.level().getServer();
        boolean whitelistEnabled = server != null && server.isEnforceWhitelist();
        String message = ArgusCore.INSTANCE.onPlayerJoin(player.getUUID(), player.hasPermissions(4), whitelistEnabled);
        if (message != null) {
            if (message.contains("Access revoked") || message.contains("Link required")) {
                player.connection.disconnect(Component.literal(message));
            } else {
                player.sendSystemMessage(Component.literal(message));
            }
        }
    }

    @SubscribeEvent
    public void onServerStarted(ServerStartedEvent event) {
        serverRef = event.getServer();
        LOGGER.info("Argus NeoForge hooks registered on server start");
        if (Boolean.getBoolean("argus.smoke")) {
            LOGGER.info("Argus NeoForge smoke flag set; stopping server after startup");
            serverRef.halt(false);
        }
    }

    @SubscribeEvent
    public void onServerStopping(ServerStoppingEvent event) {
        ArgusCore.INSTANCE.stopDiscord();
    }
}

File: neoforge/src/main/kotlin/dev/butterflysky/argus/neoforge/ArgusNeoForge.kt
package dev.butterflysky.argus.neoforge

import dev.butterflysky.argus.common.ArgusCore
import dev.butterflysky.argus.common.LoginResult
import net.minecraft.network.chat.Component
import net.neoforged.api.distmarker.Dist
import net.neoforged.fml.common.Mod
import net.neoforged.fml.event.lifecycle.FMLCommonSetupEvent
import net.neoforged.fml.javafmlmod.FMLJavaModLoadingContext
import net.neoforged.neoforge.common.NeoForge
import net.neoforged.neoforge.event.entity.player.PlayerEvent
import net.neoforged.neoforge.event.server.ServerStartedEvent
import net.neoforged.neoforge.eventbus.api.SubscribeEvent
import net.neoforged.neoforge.fml.common.Mod.EventBusSubscriber
import net.neoforged.neoforge.fml.common.Mod.EventBusSubscriber.Bus
import org.slf4j.LoggerFactory

@Mod(ArgusNeoForge.MOD_ID)
@EventBusSubscriber(modid = ArgusNeoForge.MOD_ID, bus = Bus.FORGE)
class ArgusNeoForge {
    init {
        FMLJavaModLoadingContext.get().modEventBus.addListener(this::commonSetup)
    }

    private fun commonSetup(event: FMLCommonSetupEvent) {
        ArgusCore.initialize().onFailure {
            logger.error("Failed to load Argus cache", it)
        }
        ArgusCore.startDiscord()
    }

    companion object {
        const val MOD_ID = "argus"
        private val logger = LoggerFactory.getLogger("argus-neoforge")

        @SubscribeEvent
        @JvmStatic
        fun onPlayerLogin(event: PlayerEvent.PlayerLoggedInEvent) {
            val player = event.entity
            val isOp = player.hasPermissions(4)
            val isWhitelisted = event.entity.server?.playerList?.isWhiteListed(player.gameProfile) ?: false
            when (val result = ArgusCore.onPlayerLogin(player.uuid, player.scoreboardName, isOp, isWhitelisted)) {
                LoginResult.Allow -> Unit
                is LoginResult.AllowWithKick -> player.connection.disconnect(Component.literal(result.message))
                is LoginResult.Deny -> player.connection.disconnect(Component.literal(result.message))
            }
        }

        @SubscribeEvent
        @JvmStatic
        fun onPlayerJoin(event: PlayerEvent.PlayerLoggedInEvent) {
            val player = event.entity
            ArgusCore.onPlayerJoin(player.uuid, player.hasPermissions(4))?.let {
                player.sendSystemMessage(Component.literal(it))
            }
        }

        @SubscribeEvent
        @JvmStatic
        fun onServerStarted(event: ServerStartedEvent) {
            logger.info("Argus NeoForge hooks registered on server start")
        }
    }
}
